File: 2300-successful-pairs-of-spells-and-potions/Solution.go
package Leetcode

import "sort"

func successfulPairs(spells []int, potions []int, success int64) []int {
	sort.Ints(potions)
	result := make([]int, len(spells))

	for i, spell := range spells {
		left, right := 0, len(potions)-1
		for left <= right {
			mid := left + (right-left)/2
			if int64(potions[mid])*int64(spell) >= success {
				right = mid - 1
			} else {
				left = mid + 1
			}
		}
		result[i] = len(potions) - left
	}
	return result
}

File: 2300-successful-pairs-of-spells-and-potions/Solution.cs
class Solution {
    public int[] SuccessfulPairs(int[] spells, int[] potions, long success) {
        Array.Sort(potions);
        int[] result = new int[spells.Length];
        
        for (int i = 0; i < spells.Length; i++) {
            int left = 0;
            int right = potions.Length - 1;
            while (left <= right) {
                int mid = left + (right - left) / 2;
                long product = (long) spells[i] * potions[mid];
                if (product >= success) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
            result[i] = potions.Length - left;
        }
        return result;
    }
}